### PR TITLE
[FEATURE][browser] Add bookmarks data items to the browser panel

### PR DESCRIPTION
**Pull request by nirvn**
## Description

Building on @nyalldawson 's nice bookmark manager work, this PR adds a Bookmarks node to our browser panel. Seeing it believing:
peekJ4216Z

It's about time our bookmarks become useful :wink: 

@nyalldawson , I added a few groups-related bits to your QgsBookmarkManager class. 

## Checklist

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `Fixes #11111` at the bottom of the commit message
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit
- [x] I have evaluated whether it is appropriate for this PR to be backported, backport requests are left as label or comment

### CODE REVIEW

File: python/core/auto_generated/qgsdataitem.sip.in [hunk 1]
```diff
@@ -750,6 +752,123 @@ Icon for favorites group
     virtual QVariant sortKey() const;
 
 
+};
+
+class QgsBookmarksItem : QgsDataCollectionItem
```
Comment (nyalldawson) [inline python/core/auto_generated/qgsdataitem.sip.in]: Could these be moved to app, and sit alongside the provider?

File: src/app/browser/qgsinbuiltdataitemproviders.cpp [hunk 2]
```diff
@@ -612,3 +612,52 @@ bool QgsProjectItemGuiProvider::handleDoubleClick( QgsDataItem *item, QgsDataIte
...
+    {
+      QgsCoordinateTransform ct( rect.crs(),
+                                 QgisApp::instance()->mapCanvas()->mapSettings().destinationCrs(), QgsProject::instance() );
+      try
+      {
+        canvasExtent = ct.transform( rect );
+      }
+      catch ( QgsCsException & )
+      {
+        QgisApp::instance()->messageBar()->pushWarning( tr( "Zoom to Bookmark" ), tr( "Could not reproject bookmark extent to project CRS." ) );
```
Comment (nyalldawson) [inline src/app/browser/qgsinbuiltdataitemproviders.cpp]: Strictly speaking you should get the message bar from the context, but it's all the same in app

File: src/core/qgsbrowsermodel.cpp [hunk 3]
```diff
@@ -101,6 +101,14 @@ void QgsBrowserModel::addRootItems()
     mRootItems << mFavorites;
   }
 
+  // add bookmarks
+  mBookmarks = new QgsBookmarksItem( nullptr, tr( "Bookmarks" ), QgsApplication::bookmarkManager(), QgsProject::instance()->bookmarkManager() );
```
Comment (nyalldawson) [inline src/core/qgsbrowsermodel.cpp]: Could you move this to app and implement as a provider? It'll be cleaner and will avoid the QgsProject::instance call from core

File: src/core/qgsdataitem.cpp [hunk 4]
```diff
@@ -114,6 +114,31 @@ QVariant QgsFavoritesItem::sortKey() const
...
 }
 
+QIcon QgsBookmarksItem::iconBookmarks()
+{
+  return QgsApplication::getThemeIcon( QStringLiteral( "/mActionShowBookmarks.svg" ) );
+}
+
+QVariant QgsBookmarksItem::sortKey() const
+{
+  return QStringLiteral( " 1" );
```
Comment (nyalldawson) [inline src/core/qgsdataitem.cpp]: Is the space intentional?

File: src/core/qgsdataitem.cpp [hunk 5]
```diff
@@ -1381,6 +1406,89 @@ QVector<QgsDataItem *> QgsFavoritesItem::createChildren( const QString &favDir,
...
+{
+  mType = BookmarkManager;
+  mManager = manager;
+  mIconName = QStringLiteral( "/mIconFolder.svg" );
+  populate();
+}
+
+QVector<QgsDataItem *> QgsBookmarkManagerItem::createChildren()
+{
+  connect( mManager, &QgsBookmarkManager::bookmarkAdded, [ = ]( const QString & ) { refresh(); } );
```
Comment (nyalldawson) [inline src/core/qgsdataitem.cpp]: Why not in constructor?

File: src/core/qgsdataitem.cpp [hunk 6]
```diff
@@ -1381,6 +1406,89 @@ QVector<QgsDataItem *> QgsFavoritesItem::createChildren( const QString &favDir,
...
+    children << new QgsBookmarkItem( this, bookmark.name(), bookmark );
+  }
+  return children;
+}
+
+QgsBookmarkItem::QgsBookmarkItem( QgsDataItem *parent, const QString &name, QgsBookmark bookmark )
+  : QgsDataItem( Bookmark, parent, name, QStringLiteral( "bookmarks:%1/%2" ).arg( bookmark.group().toLower(), bookmark.id() ) )
+{
+  mType = Bookmark;
+  mCapabilities |= Fertile;
```
Comment (nyalldawson) [inline src/core/qgsdataitem.cpp]: Hang on, we don't want the fertile flag here - that indicates that the item can have children

**Review by nyalldawson — COMMENTED** [review]
I think youll need to use the "Fast" flag for the new items -- these will need to be populated in the main thread as bookmark manager isn't thread safe.

File: src/core/qgsdataitem.cpp (hunk 6 above)
Comment (nirvn) [reply src/core/qgsdataitem.cpp]: Am I not removing the Fertile flag here?

File: src/core/qgsdataitem.cpp (hunk 5 above)
Comment (nirvn) [reply src/core/qgsdataitem.cpp]: True.

File: src/core/qgsdataitem.cpp (hunk 4 above)
Comment (nirvn) [reply src/core/qgsdataitem.cpp]: Yes, to insure the item is placed below favorites.

File: src/core/qgsdataitem.cpp (hunk 6 above)
Comment (nyalldawson) [reply src/core/qgsdataitem.cpp]: Nope, you're adding it ;)

File: src/app/qgsappbrowserproviders.cpp [hunk 7]
```diff
@@ -527,3 +557,138 @@ QgsDataItem *QgsProjectDataItemProvider::createDataItem( const QString &path, Qg
...
+//
+
+QString QgsBookmarksDataItemProvider::name()
+{
+  return QStringLiteral( "bookmarks_item" );
+}
+
+int QgsBookmarksDataItemProvider::capabilities() const
+{
+  return QgsDataProvider::File;
```
Comment (nyalldawson) [inline src/app/qgsappbrowserproviders.cpp]: But this one shouldn't be file....

File: src/app/qgsappbrowserproviders.cpp [hunk 8]
```diff
@@ -527,3 +557,138 @@ QgsDataItem *QgsProjectDataItemProvider::createDataItem( const QString &path, Qg
...
+}
+
+QVector<QgsDataItem *> QgsBookmarkManagerItem::createChildren()
+{
+  QVector<QgsDataItem *> children;
+  for ( const QString &group : mManager->groups() )
+  {
+    if ( group.isEmpty() )
+    {
+      for ( const QgsBookmark bookmark : mManager->bookmarksByGroup( QString( "" ) ) )
```
Comment (nyalldawson) [inline src/app/qgsappbrowserproviders.cpp]:
```suggestion
      for ( const QgsBookmark bookmark : mManager->bookmarksByGroup( QString() ) )
```

File: src/app/qgsappbrowserproviders.cpp [hunk 9]
```diff
@@ -515,7 +545,7 @@ QString QgsProjectDataItemProvider::name()
 
 int QgsProjectDataItemProvider::capabilities() const
 {
-  return QgsDataProvider::File;
+  return QgsDataProvider::Database;
```
Comment (nyalldawson) [inline src/app/qgsappbrowserproviders.cpp]: This still needs to be file, please don't change.

**Comment by nirvn** [comment]
@nyalldawson , all done; I've also implemented a right click menu item for the parent bookmarks node. This feels real good now.

**Pull request merged**